### PR TITLE
plan: Support “in” and “like” in new check

### PR DESCRIPTION
**Pull request by zimulala**

### CODE REVIEW

**Comment by zimulala** [comment]
@hanfei19910905 @coocood @shenli PTAL

File: plan/refiner.go [hunk 1]
```diff
@@ -437,7 +436,43 @@ func (c *conditionChecker) checkScalarFunction(scalar *expression.ScalarFunction
...
+			}
+		}
 		return c.newCheck(scalar.Args[0])
+	case ast.In:
+		if !c.checkColumn(scalar.Args[0]) {
+			return false
+		}
+		for _, v := range scalar.Args[1:] {
+			if _, ok := v.(*expression.Constant); !ok {
+				return false
```
Comment (hanfei19910905) [inline plan/refiner.go]: If a in (1,2,b,3) , it can be treated as a in (1,2,3) or a = b

File: plan/refiner.go [hunk 2]
```diff
@@ -437,7 +436,43 @@ func (c *conditionChecker) checkScalarFunction(scalar *expression.ScalarFunction
...
+		}
+		patternStr, err := pattern.Value.ToString()
+		if err != nil {
+			return false
+		}
+		if len(patternStr) == 0 {
+			return true
+		}
+		firstChar := patternStr[0]
+		return firstChar != '%' && firstChar != '.'
```
Comment (shenli) [inline plan/refiner.go]: select a from t where c like 'abc%de'
The like expr could be converted to accesss condition. But it should also be added into filter condition.

**Comment by zimulala** [comment]
@shenli PTAL

File: plan/refiner.go [hunk 3]
```diff
@@ -437,11 +444,65 @@ func (c *conditionChecker) checkScalarFunction(scalar *expression.ScalarFunction
...
+	}
+	patternStr, err := pattern.Value.ToString()
+	if err != nil {
+		return false
+	}
+	if len(patternStr) == 0 {
+		return true
+	}
+	for i := 0; i < len(patternStr); i++ {
+		if i == 0 && (patternStr[i] == '%' || patternStr[i] == '_') {
```
Comment (shenli) [inline plan/refiner.go]: Consider escape syntax in pattern string.

File: plan/refiner.go [hunk 4]
```diff
@@ -437,11 +444,65 @@ func (c *conditionChecker) checkScalarFunction(scalar *expression.ScalarFunction
 	case ast.IsNull, ast.IsTruth, ast.IsFalsity:
 		return c.checkColumn(scalar.Args[0])
 	case ast.UnaryNot:
+		// Don't support "not like" and "not in" convert to access conditions.
+		if s, ok := scalar.Args[0].(*expression.ScalarFunction); ok {
```
Comment (coocood) [inline plan/refiner.go]: What if `s` is OR expression, e.g. `select * from x where not (a like "a" or a like "ab")`?

File: plan/refiner.go [hunk 5]
```diff
@@ -309,11 +318,11 @@ func buildNewTableRange(p *PhysicalTableScan) error {
...
 type conditionChecker struct {
-	tableName model.CIStr
-	idx       *model.IndexInfo
-	// the offset of the indexed column to be checked.
-	columnOffset int
+	tableName    model.CIStr
+	idx          *model.IndexInfo
+	columnOffset int // the offset of the indexed column to be checked.
 	pkName       model.CIStr
+	isReserved   bool // check if a access condition can be reserved to filter conditions.
```
Comment (coocood) [inline plan/refiner.go]: isReserved -> shouldReserve
can be reserved to -> should be reserved in

**Comment by zimulala** [comment]
@coocood @shenli @hanfei19910905 PTAL

**Comment by coocood** [comment]
LGTM

**Comment by hanfei19910905** [comment]
open the test : https://github.com/pingcap/tidb/blob/master/session_test.go#L2184

**Comment by shenli** [comment]
LGTM

**Comment by zimulala** [comment]
@hanfei19910905 PTAL

**Comment by hanfei19910905** [comment]
LGTM

**Pull request merged**